### PR TITLE
feat(eval): synthetic mbox fixture for email-agent eval (#848)

### LINKED ISSUE #848: feat(email): synthetic .mbox dataset for email triage agent testing
## Summary

The Email Triage Agent (Phase 2, v0.23.0) needs a synthetic `.mbox` dataset to validate the triage engine, MIME parsing, email threading, and categorization logic without requiring real email accounts or OAuth setup. This unblocks offline development and CI testing of the email pipeline described in the [email & calendar integration plan](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx).

## Why

- **No real inboxes in CI** — tests cannot depend on live IMAP/OAuth credentials
- **Reproducible triage evaluation** — ground-truth labels let us measure categorization accuracy against the [80% target](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx#12-success-metrics)
- **MIME edge cases** — multipart, attachments, inline images, non-UTF-8 encodings, and nested forwarding need coverage upfront
- **Threading correctness** — `In-Reply-To` / `References` chains must parse correctly for conversation grouping

## Proposed Dataset

Generate a synthetic `.mbox` file (Python `mailbox` + `email` stdlib) with **~200-250 messages** covering a realistic inbox mix.

### Triage categories (ground truth labels in metadata)

The plan (section 4.2) defines 4 triage categories. This dataset adds a **pre-triage spam/phishing class** — the triage engine should filter these *before* categorization, not treat them as a 5th category. Ground truth marks them with `is_spam: true` / `is_phishing: true` separately from `category`.

| Category | Count | Examples |
|----------|-------|---------|
| **Urgent** | ~20 | Boss escalation, client contract deadline, prod incident alert, security advisory |
| **Actionable** | ~45 | PR review requests, meeting invites, direct questions, expense approvals |
| **Informational** | ~55 | Newsletters, order receipts, shipping notifications, internal announcements |
| **Low priority** | ~30 | Marketing promos, social media notifications, cold outreach |
| **Spam/phishing** (pre-triage filter) | ~20 | Phishing attempts, fake invoices, scams, fake delivery notices |
| **Ambiguous/borderline** | ~15 | Hard-to-classify messages (see section below) |

Messages within each category are drawn from the corporate, personal, and spam flavors described below — the flavors are **content types within categories**, not separate pools.

### Recurring sender personas

Define 8-10 recurring senders so threading and sender-importance learning can be tested across multiple messages from the same person:

| Persona | Domain | Role | Typical category |
|---------|--------|------|-----------------|
| Sarah Chen (boss) | `@acme-corp.example.com` | VP Engineering | Urgent / Actionable |
| Alex Kumar (direct report) | `@acme-corp.example.com` | Senior Engineer | Actionable |
| Jordan Lee (cross-team) | `@acme-corp.example.com` | Product Manager | Actionable / Informational |
| IT Systems | `noreply@acme-corp.example.com` | Automated | Informational |
| HR Team | `hr@acme-corp.example.com` | Automated | Informational |
| Maria Santos (client) | `@globaltech.example.net` | External partner | Urgent / Actionable |
| DevOps Bot | `alerts@acme-corp.example.com` | CI/CD / PagerDuty | Urgent / Informational |
| Newsletter senders | various `@*.example.com` | Marketing / news | Informational / Low priority |

Each recurring sender should appear in **3-8 messages** (mix of thread roots and replies) to enable sender-frequency and response-pattern analysis.

### Corporate / enterprise email types (~60-70 messages)

The primary target is enterprise users on Outlook/Exchange. These messages are distributed across the triage categories above:

- **IT / Ops notifications** — password expiry warnings, VPN maintenance windows, system outage alerts, new software rollout announcements
- **HR / People Ops** — benefits enrollment reminders, PTO policy updates, org announcements, new hire intros, mandatory training deadlines
- **Executive comms** — all-hands meeting recap, quarterly earnings summary, CEO update, reorg announcement
- **Cross-team threads** — multi-person `Cc` chains with 4-6 participants, "loop in [name]" forwards, top-posted reply style
- **Calendar-related** — meeting invite (`.ics`), meeting cancellation, room booking confirmation, recurring meeting update
- **Compliance / Legal** — NDA reminder, export control notice, data retention policy update, audit request
- **Automated systems** — JIRA ticket assignment, CI/CD build failure, Confluence page update, PagerDuty alert, Salesforce lead notification
- **Expense / Finance** — expense report approval, purchase order confirmation, travel booking receipt, budget review request

### Spam / phishing (~20 messages, pre-triage filter)

Realistic spam that the triage engine should filter before categorization:

- **Phishing** — fake password reset, "verify your account", spoofed IT department, fake DocuSign
- **Scam** — Nigerian prince, lottery winner, inheritance notification, crypto "opportunity"
- **Commercial spam** — unsolicited product pitch, SEO services, fake invoice attachment, "limited time offer"
- **Social engineering** — fake LinkedIn connection, spoofed coworker name with external domain, urgent wire transfer request from "CEO"
- **Delivery scams** — fake UPS/FedEx tracking, "package held at customs", Amazon order you did not place

Spam messages should include realistic spam signals: mismatched `From` display name vs address, suspicious domains, urgency language, misspellings, suspicious attachment names, missing `List-Unsubscribe`, etc.

### Personal / consumer email types (~30 messages)

- Order confirmations and shipping updates (Amazon, retailer receipts)
- Subscription newsletters (tech blogs, news digests)
- Social media notifications (LinkedIn, GitHub stars/follows)
- Bank/financial alerts (transaction notification, statement ready)
- Travel confirmations (flight, hotel, car rental)
- App notifications (Slack digest, calendar reminders)

### Ambiguous / borderline messages (~15 messages)

Intentionally hard-to-classify emails that stress-test the triage engine decision boundaries. Mark these with `ambiguous: true` in ground truth, plus a `rationale` field explaining the intended classification:

- Meeting invite from unknown external contact (Actionable or Low priority?)
- Vendor invoice with no prior relationship (Urgent or Informational?)
- Automated JIRA ticket from a project the user is not on (Actionable or Informational?)
- Newsletter from a tool you actively use vs. one signed up for and forgotten (Informational or Low priority?)
- "Quick question" from someone never emailed before (Actionable or Low priority?)
- Internal compliance email that requires acknowledgment by EOD (Urgent or Actionable?)
- Reply-all on a thread where the user was Cc'd but not addressed (Actionable or Informational?)

### Malformed / edge-case messages (~5-10 messages)

Real inboxes contain broken email. Include parser-robustness tests:

- Missing `Subject` header
- Empty body (headers only)
- Truncated multipart (missing closing boundary)
- Invalid `Date` header (unparseable format)
- Double-encoded UTF-8 subject (`=?UTF-8?B?...?=` wrapping already-encoded text)
- Base64-encoded body with incorrect padding
- Extremely long `Subject` (>500 chars)
- Message with no `From` header

### Email structure variety

- Plain text only
- HTML only (with realistic corporate email templates)
- Multipart (text + HTML)
- With attachments (small dummy `.pdf`, `.csv`, `.png`, `.docx` — **1-5 KB each, total .mbox under 1 MB**)
- Inline images (`Content-Disposition: inline`)
- Forwarded messages (nested `message/rfc822`)
- Reply chains (3-5 deep with proper `In-Reply-To` / `References` headers)
- Non-ASCII subjects and bodies (UTF-8, ISO-8859-1)
- Calendar invites (`.ics` attachments)
- Top-posted replies with `>` quoted original (Outlook style)
- HTML signature blocks with logos and legal disclaimers

### Realistic metadata

- Varied `From` addresses using the recurring sender personas above
- Corporate domains: `@acme-corp.example.com`, `@globaltech.example.net` (RFC 2606 reserved)
- Realistic `Date` headers spanning ~2 weeks, **clustered around 9 AM and 4-5 PM** on weekdays to simulate real arrival patterns, with a batch arriving overnight Saturday to Monday for "overnight triage" testing
- Proper `Message-ID`, `In-Reply-To`, `References` for threading
- Mix of `To`, `Cc`, `Bcc` patterns (including large `Cc` lists for corporate threads)
- `List-Unsubscribe` headers on marketing/newsletter emails
- `X-Priority` / `Importance` headers on some urgent emails
- `X-Mailer` headers (Outlook, Thunderbird, Gmail, automated systems)
- Corporate email disclaimers in footers ("This email is confidential...")
- `Reply-To` mismatches on phishing emails
- `Received` header chains simulating realistic relay paths

## Deliverables

1. **Generator script** — `tests/fixtures/email/generate_mbox.py` that produces the `.mbox` deterministically (seeded RNG)
2. **Pre-built fixture** — `tests/fixtures/email/synthetic_inbox.mbox` checked into repo (must be under 1 MB)
3. **Ground truth manifest** — `tests/fixtures/email/ground_truth.json` mapping `Message-ID` to `{ category, priority, is_thread_root, thread_id, has_attachment, is_spam, is_phishing, ambiguous, rationale, sender_persona }`
4. **Pytest fixtures** — in `tests/fixtures/email/conftest.py`:
   - `synthetic_mbox` — loads the `.mbox` file
   - `email_ground_truth` — loads the ground truth JSON
   - `single_email(category)` — returns one message of a given triage category
   - `spam_emails` — returns all spam/phishing messages for filter testing
   - `ambiguous_emails` — returns borderline messages for boundary testing

## Acceptance criteria

- [ ] `python tests/fixtures/email/generate_mbox.py` produces a valid `.mbox` readable by Python `mailbox.mbox()`
- [ ] All 4 triage categories represented with ground truth labels, plus separate `is_spam`/`is_phishing` flags (spam is a pre-triage filter, not a 5th category)
- [ ] Corporate email types cover IT, HR, exec comms, cross-team threads, and automated systems
- [ ] Spam/phishing emails include realistic indicators (domain mismatches, urgency, suspicious attachments)
- [ ] ~15 ambiguous/borderline messages included with `ambiguous: true` and `rationale` in ground truth
- [ ] ~5-10 malformed messages test parser robustness (missing headers, truncated MIME, encoding errors)
- [ ] Recurring sender personas appear in 3-8 messages each for sender-importance testing
- [ ] Multipart, attachment, threading, and encoding edge cases covered
- [ ] No real PII — all names, addresses, domains, and content are synthetic (use `.example.com` / `.example.net`)
- [ ] Total `.mbox` file size under 1 MB; individual dummy attachments 1-5 KB each
- [ ] Pre-built fixture matches generator output (hash check in CI — `python tests/fixtures/email/generate_mbox.py --verify`)
- [ ] Pytest can load and iterate the dataset (`tests/unit/test_synthetic_mbox.py` passes)
- [ ] Generator is deterministic (same seed produces identical output)

## References

- Plan: [`docs/plans/email-calendar-integration.mdx`](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx) section 4 (Email Triage Agent)
- Related: #645 (Email Triage Agent), #663 (Daily briefs)

### PR DESCRIPTION
**Pull request by theonlychant**
## Summary
Adds a synthetic `.mbox` dataset for email triage agent evaluation, with 
a deterministic generator that writes fixtures to a temporary directory 
at test time so the source tree is never mutated.

## Why
The email triage agent had no test data to validate triage logic in CI. 
This fixture provides realistic email threads covering all required 
categories, edge cases, and threading headers without requiring a live 
mailbox.

## Linked issue
Closes #848

## Changes
- Updated test fixtures to use `tests/fixtures/email/ground_truth.json` 
  if it already exists; otherwise the generator writes deterministic 
  `ground_truth.json` and `synthetic_inbox.mbox` into a temporary 
  directory at test time - source tree is never mutated
- Fixed `test_generator_determinism_verify_mode` to call 
  `_ensure_generated()` itself so it's no longer order-dependent
- Tightened filesystem and browser keyword regexes to use word boundaries 
  - prevents mis-routing on slashes in URLs, "and/or", "online meeting"
- Added `force_activate(bundle_name)` public method to `ToolLoader` to 
  replace direct `_state` access
- Wired `reset_session()` into conversation-start path

## Test plan
- [x] `pytest tests/unit/test_synthetic_mbox.py` - all passing
- [x] `pytest tests/unit/test_synthetic_mbox.py::test_generator_determinism_verify_mode` - passes in isolation
- [x] `python util/lint.py --all` - no failures

## Checklist
- [x] I have linked a GitHub issue above (`Closes #848`).
- [x] I have described **why** this change is being made, not just what changed.
- [x] I have run linting and tests locally.
- [x] Documentation updated where applicable.

### CODE REVIEW

**Comment by theonlychant** [comment]
for some reason the pylint keeps running into problems on my tests saying 
```bash
python3 util/lint.py --pylint
========================================
Running Code Quality Checks
========================================

[3/9] Running Pylint (errors only)...
----------------------------------------
[CMD] pylint src/gaia --rcfile .pylintrc --disable C0103,C0301,W0246,W0221,E1102,R0401,E0401,W0718,W0212

[!] Pylint found critical errors:
************* Module gaia.agents.base.console
src/gaia/agents/base/console.py:909:46: E1101: Module 'PIL.Image' has no 'LANCZOS' member (no-member)
************* Module gaia.mcp.servers.agent_ui_mcp
src/gaia/mcp/servers/agent_ui_mcp.py:441:43: E1101: Module 'PIL.Image' has no 'LANCZOS' member (no-member)



================================================================
                    LINT SUMMARY REPORT                        
================================================================

[STATS] Project Statistics:
   - Python Files: 355
   - Lines of Code: 160,254
   - Directories: src/gaia, tests

[RESULTS] Quality Check Results:

+--------------------------------+------------+-----------+
| Check                          | Status     | Issues    |
+--------------------------------+------------+-----------+
| Critical Errors (Pylint)       | [X] FAIL   | 2 errors  |
+--------------------------------+------------+-----------+

[SUMMARY] Statistics:
   - Total Checks Run: 1
   - Passed: 0 (0.0%)
   - Failed: 1 (100.0%)
   - Warnings: 0 (0.0%)

============================================================
[FAILED] QUALITY CHECKS FAILED
============================================================

[ERROR] Issues Found:
   - 1 critical error(s) - must fix before PR

[TIP] Review the error messages above and fix the issues.
[TIP] Use --fix flag to auto-fix formatting issues:
   python util/lint.py --black --fix
   python util/lint.py --isort --fix
```
I tried resolving the issue nothing seems to work right now I will try to fix it later 

I made some changes in a different branch [PR](https://github.com/theonlychant/gaia/commit/ebbc78412d03beb372191ed70c82a177eb46aeed)

**Comment by theonlychant** [comment]
I ran this again in my kernel the checks were good 
```bash
================================================================
                    LINT SUMMARY REPORT                        
================================================================

[STATS] Project Statistics:
   - Python Files: 362
   - Lines of Code: 162,969
   - Directories: src/gaia, tests

[RESULTS] Quality Check Results:

+--------------------------------+------------+-----------+
| Check                          | Status     | Issues    |
+--------------------------------+------------+-----------+
| Code Formatting (Black)        | [OK] PASS  | -         |
| Import Sorting (isort)         | [OK] PASS  | -         |
| Critical Errors (Pylint)       | [OK] PASS  | -         |
| Style Compliance (Flake8)      | [OK] PASS  | -         |
| Type Checking (MyPy)           | [!] WARN   | 1109 warns |
| Import Validation              | [OK] PASS  | -         |
| Security Check (Bandit)        | [!] WARN   | 54 warns  |
| Agent Conventions              | [!] WARN   | 11 warns  |
| Doc Version Consistency        | [OK] PASS  | -         |
+--------------------------------+------------+-----------+

[SUMMARY] Statistics:
   - Total Checks Run: 9
   - Passed: 6 (66.7%)
   - Failed: 0 (0.0%)
   - Warnings: 3 (33.3%)

============================================================
[SUCCESS] ALL QUALITY CHECKS PASSED!
[WARNING] 3 warning(s) found (non-blocking)
============================================================

[OK] Your code meets quality standards!
[OK] Ready for PR submission
```

File: tests/fixtures/email/conftest.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+EMAIL_FIXTURE_DIR = Path(__file__).resolve().parent
+MBOX_PATH = EMAIL_FIXTURE_DIR / "synthetic_inbox.mbox"
+GT_PATH = EMAIL_FIXTURE_DIR / "ground_truth.json"
+
+
+def _ensure_generated() -> None:
+    if MBOX_PATH.exists() and GT_PATH.exists():
+        return
+    generate(MBOX_PATH, GT_PATH)
```
Comment (itomek-amd) [inline tests/fixtures/email/conftest.py]: **Blocking.** Issue #848 asks for `synthetic_inbox.mbox` and `ground_truth.json` to be committed as pre-built fixtures, but they're not in the PR — so this fallback writes ~1MB into the source tree on first test run. Every contributor and CI shard ends up with a dirty working copy on import. Commit the generated artifacts, or write to `tmp_path` instead.

File: tests/unit/test_synthetic_mbox.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    for message_id, meta in gt.items():
+        assert message_id.startswith("<") and message_id.endswith(">")
+        assert required.issubset(meta.keys())
+
+
+def test_generator_determinism_verify_mode() -> None:
+    cmd = [sys.executable, str(GEN_SCRIPT), "--verify"]
+    proc = subprocess.run(cmd, capture_output=True, text=True, check=False)
+    assert proc.returncode == 0, proc.stdout + "\n" + proc.stderr
```
Comment (itomek-amd) [inline tests/unit/test_synthetic_mbox.py]: **Blocking — flaky.** `generate_mbox.py --verify` returns 1 when the fixture files don't exist ("Missing pre-built fixtures; run without --verify first."). This test only passes today because earlier tests in the module trigger `_ensure_generated()` first. Run it in isolation and it fails. Either call `_ensure_generated()` from this test, or generate into `tmp_path` and verify against that.

File: src/gaia/agents/chat/agent.py [hunk 3]
```diff
@@ -1704,6 +1709,241 @@ def search_web(query: str) -> dict:
...
+                        "write_file",
+                        "edit_file",
+                    }
+                ),
+                policy=ActivationPolicy.KEYWORD,
+                keywords=frozenset(
+                    {
+                        r"file|folder|directory|path|find|search|browse|tree|ls\b|dir\b",
+                        r"\.[a-z]{1,5}\b",  # file extensions like .py, .json
+                        r"[/\\]",  # path separators
```
Comment (itomek-amd) [inline src/gaia/agents/chat/agent.py]: **Blocking.** `r"[/\\]"` matches *any* slash anywhere in the user message — so `"https://example.com"`, `"and/or"`, `"24/7"`, and any error message containing a path will activate filesystem tools. Use `r"(?:^|\s)[\./~\\][\w/\\.-]*"` or similar so a bare slash inside a URL doesn't trigger.

File: src/gaia/agents/chat/agent.py [hunk 4]
```diff
@@ -1704,6 +1709,241 @@ def search_web(query: str) -> dict:
...
+                        "search_web",
+                        "search_documentation",
+                    }
+                ),
+                policy=ActivationPolicy.KEYWORD,
+                keywords=frozenset(
+                    {
+                        r"https?://",
+                        r"url|website|webpage|web\s*page|browse|internet",
+                        r"search.*web|google|look\s*up|online",
```
Comment (itomek-amd) [inline src/gaia/agents/chat/agent.py]: **Blocking.** This activates browser tools on `"look it up in the doc"` (matches `look\s*up`), `"online meeting"` and `"online course"` (matches `online`). Tighten to word boundaries: `r"\bgoogle\s"`, `r"\blook\s+(?:it\s+)?up\s+(?:online|on the web)\b"`, `r"\bonline\s+(?:search|article|forum)\b"`.

File: src/gaia/agents/base/tool_loader.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+    def get_active_bundle_names(self) -> list[str]:
+        """Return names of currently activated bundles."""
+        return [n for n, s in self._state.items() if s.activated]
+
+    def get_bundle_for_tool(self, tool_name: str) -> Optional[str]:
+        """Return the bundle name that owns *tool_name*, or ``None``."""
+        return self._tool_to_bundle.get(tool_name)
+
+    def reset_session(self) -> None:
```
Comment (itomek-amd) [inline src/gaia/agents/base/tool_loader.py:240]: Non-blocking. `reset_session()` isn't called from anywhere in this PR — grep'd `src/` and only the unit test references it. Either wire it into the conversation-start path (e.g. `Agent.reset_state()`) or drop it; dead public API tends to rot.

**Review by itomek-amd — CHANGES_REQUESTED** [review]
Thanks for bundling this — there's real, useful work here (the `ToolLoader` design is clean, the bundle policies are well-thought-out, and the synthetic-mbox generator is impressively thorough). Requesting changes on a few items before merge.

_Findings below are from a code read; I didn't run pytest or lint locally, so the tests may pass cleanly. Disregard the existing PR-check failure (`LANCZOS` pylint error in `screenshot_tools.py`) — it reproduces on `main` and is pre-existing tech debt, not from this PR._

## Blocking

1. **Pre-built fixtures (`synthetic_inbox.mbox`, `ground_truth.json`) are missing from the repo.** Issue #848 explicitly asks for them to be committed. Without them, `tests/fixtures/email/conftest.py` `_ensure_generated()` silently generates a ~1MB `.mbox` into the source tree on first test run — meaning every contributor and CI runner mutates their working copy on import. Either commit the generated artifacts, or move generation to `tmp_path` so the source tree stays clean.

2. **`test_generator_determinism_verify_mode` is order-dependent.** It runs `generate_mbox.py --verify`, but if the fixture files aren't present yet the script returns `1` with `"Missing pre-built fixtures; run without --verify first."`. The test only passes today because earlier tests in the module call `_ensure_generated()` first. Run this test in isolation (`pytest tests/unit/test_synthetic_mbox.py::test_generator_determinism_verify_mode`) and it will fail. Fix: ensure the `--verify` test calls `_ensure_generated()` itself, or test against a `tmp_path`-generated copy.

3. **Two keyword regexes in `_setup_tool_bundles` are too greedy and will mis-route.**
   - `filesystem` includes `r"[/\\]"` — matches *any* slash, so `"https://example.com"`, `"and/or"`, `"24/7"`, code paths in error messages all activate filesystem tools.
   - `browser` includes `r"search.*web|google|look\s*up|online"` — matches `"look it up in the doc"`, `"online meeting"`, and any sentence containing the substring `online`.
   Tighten to word-boundary patterns (e.g. `r"\bhttps?://"`, `r"\bgoogle\s"`, `r"\blook\s+up\b"`).

## Non-blocking

4. `ToolLoader.reset_session()` is defined but never called — confirm it's wired into the conversation-start path or drop it.
5. `ChatAgent._setup_tool_bundles` writes directly to `self.tool_loader._state["rag"].activated = True`. Add a public `force_activate(bundle_name)` method on the loader so this isn't an encapsulation breach.
6. `Agent.rebuild_system_prompt()` builds against the full `_TOOL_REGISTRY` rather than passing through the loader — verify this is intentional (it means the loader's filter only applies on the initial prompt build, not on rebuild).
7. PR title `"feat: PR for issue #848"` doesn't follow the repo's conventional-commits convention. Consider `feat(eval): synthetic mbox fixture for email-agent eval (#848)`.

Once these land I'm happy to re-review and approve.
